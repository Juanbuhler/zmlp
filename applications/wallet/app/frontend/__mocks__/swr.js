import { createElement } from 'react'

const noop = () => () => {}

/**
 * <SWRConfig />
 */

export const SWRConfig = ({ children, ...rest }) =>
  createElement('SWRConfig', rest, children)

/**
 * useSWR
 */

let mockUseSWRResponse = {}

export const __setMockUseSWRResponse = data => {
  mockUseSWRResponse = { revalidate: noop, ...data }
}

const useSWR = () => {
  return mockUseSWRResponse
}

export default useSWR
