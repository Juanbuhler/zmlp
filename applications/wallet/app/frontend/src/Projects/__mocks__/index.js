import { createElement } from 'react'

const Projects = ({ children, ...rest }) =>
  createElement('Projects', rest, children(rest))

export const projects = [
  {
    url: 'http://localhost:3000/api/v1/projects/1',
    name: 'Zorroa ABC',
  },
  {
    url: 'http://localhost:3000/api/v1/projects/2',
    name: 'Zorroa EasyAs123',
  },
  {
    url: 'http://localhost:3000/api/v1/projects/3',
    name: 'Zorroa JustYouAndMe',
  },
]

export default Projects
