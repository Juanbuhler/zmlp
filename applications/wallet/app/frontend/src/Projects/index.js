import PropTypes from 'prop-types'
import useSWR from 'swr'

import Layout from '../Layout'

const Projects = ({ logout, children }) => {
  const { data: { results = [] } = {} } = useSWR('api/v1/projects/')

  if (results.length === 0) return 'Loading...'

  return (
    <Layout logout={logout} results={results}>
      {children}
    </Layout>
  )
}

Projects.propTypes = {
  logout: PropTypes.func.isRequired,
  children: PropTypes.func.isRequired,
}

export default Projects
