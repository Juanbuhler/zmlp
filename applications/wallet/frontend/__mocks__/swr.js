import { createElement } from 'react'

const noop = () => () => {}

/**
 * <SWRConfig />
 */

export const SWRConfig = ({ children, ...rest }) =>
  createElement('SWRConfig', rest, children)

/**
 * cache
 */

export const { cache } = jest.requireActual('swr')

/**
 * useSWR
 */

let mockUseSWRResponse = { revalidate: noop, mutate: noop }

export const __setMockUseSWRResponse = (data) => {
  mockUseSWRResponse = { revalidate: noop, mutate: noop, ...data }
}

const useSWR = () => {
  return mockUseSWRResponse
}

export default useSWR
