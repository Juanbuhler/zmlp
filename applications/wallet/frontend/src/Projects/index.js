import { useEffect } from 'react'
import PropTypes from 'prop-types'
import Router, { useRouter } from 'next/router'
import useSWR from 'swr'

const NO_PROJECT_ID_ROUTES = ['/account', '/account/password']

const Projects = ({ projectId, setUser, children }) => {
  const { query: { projectId: routerProjectId } = {}, pathname } = useRouter()

  const {
    data: { results: projects },
  } = useSWR('/api/v1/projects/')

  useEffect(() => {
    if (!routerProjectId) return

    if (projectId === routerProjectId) return

    setUser({ user: { projectId: routerProjectId } })
  }, [projectId, routerProjectId, setUser])

  // Reset user projectId if not part of current projects
  if (projectId && !projects.find(({ id }) => projectId === id)) {
    setUser({ user: { projectId: '' } })
    return null
  }

  // Render "No Projects" page
  if (projects.length === 0 && pathname === '/') {
    return children
  }

  // Render other pages without projectId
  if (!routerProjectId && NO_PROJECT_ID_ROUTES.includes(pathname)) {
    return children
  }

  // Redirect pages that require a projectId to "No Projects" page if there are no projects
  if (projects.length === 0 && !NO_PROJECT_ID_ROUTES.includes(pathname)) {
    Router.push('/')
    return null
  }

  // Remove url projectId if not part of current projects
  if (!projects.find(({ id }) => routerProjectId === id)) {
    Router.push('/[projectId]/jobs', `/${projectId || projects[0].id}/jobs`)
    return null
  }

  return children
}

Projects.propTypes = {
  projectId: PropTypes.string.isRequired,
  setUser: PropTypes.func.isRequired,
  children: PropTypes.node.isRequired,
}

export default Projects
