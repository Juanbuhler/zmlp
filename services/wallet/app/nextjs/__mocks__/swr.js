import { createElement } from 'react'

/**
 * <SWRConfig />
 */

export const SWRConfig = ({ children, ...rest }) =>
  createElement('SWRConfig', rest, children)

/**
 * useSWR
 */

let mockUseSWRResponse = {}

export const __setMockUseSWRResponse = data => {
  mockUseSWRResponse = data
}

const useSWR = () => {
  return mockUseSWRResponse
}

export default useSWR
