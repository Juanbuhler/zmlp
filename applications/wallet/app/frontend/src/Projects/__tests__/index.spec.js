import TestRenderer from 'react-test-renderer'

import Projects from '..'

jest.mock('../../Layout')

describe('<Projects />', () => {
  it('should render properly without data', () => {
    const mockFn = jest.fn()

    const component = TestRenderer.create(
      <Projects logout={mockFn}>{() => `Hello World`}</Projects>,
    )

    expect(component.toJSON()).toMatchSnapshot()
  })

  it('should render properly with data', () => {
    const mockFn = jest.fn()
    require('swr').__setMockUseSWRResponse({
      data: { results: [{ id: '1', name: 'project-name' }] },
    })

    const component = TestRenderer.create(
      <Projects logout={mockFn}>{() => `Hello World`}</Projects>,
    )

    expect(component.toJSON()).toMatchSnapshot()
  })
})
