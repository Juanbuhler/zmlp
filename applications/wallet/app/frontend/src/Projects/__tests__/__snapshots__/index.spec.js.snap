// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Projects /> should render properly with data 1`] = `
<Layout
  logout={[MockFunction]}
  results={
    Array [
      Object {
        "id": "1",
        "name": "project-name",
      },
    ]
  }
>
  Hello World
</Layout>
`;

exports[`<Projects /> should render properly without data 1`] = `"Loading..."`;
